mesonbuild/msetup.py

```diff
@@ -160,6 +160,10 @@ def add_vcs_ignore_files(self, build_dir: str) -> None:
 
     def validate_dirs(self, dir1: str, dir2: str, reconfigure: bool, wipe: bool) -> T.Tuple[str, str]:
         (src_dir, build_dir) = self.validate_core_dirs(dir1, dir2)
+
+        if os.path.pardir not in os.path.relpath(src_dir, build_dir):
+            raise SystemExit(f'builddir {build_dir} cannot be a parent of sourcedir {src_dir}')
```

Comment:
I would use os.path.commonpath here, and simply check if the common path is the build directory.

Reply:
Oh also it seems the Windows tests *really* don't like you using relpath on multiple drives. :)

Reply:
You are right! Normally, I would use Path.is_relative for that, but it is only available in Python 3.9...

Reply:
Using commonpath should mean you don't need to handle ValueError at all, ntpath.commonpath doesn't raise one here.

That's why I said that it's specifically using relpath that causes drive issues.

Reply:
Are-you sure?

> Raise ValueError if paths contain both absolute and relative pathnames, the paths are on the different drives or if paths is empty.

Reply:
Hmm, I could swear I tried it last night and it worked. But maybe I'm losing my mind.

Reply:
`os.path.commonprefix` doesn't throw, but `os.path.commonpath` does. But we really want `commonpath` here, and the cost of exception handling should be neglectable in this context.

Reply:
I agree, but I *thought* I tested commonpath, not commonprefix. Again, I may have simply been losing my mind. :D